gitpod-io/gitpod
# Improve Modal a11y

## Description

A few changes to improve our `<Modal/>` component's accessibility. Needed a library, to handle focus locking when a modal is open, and turns out it (`react-focus-on`) handles a few of the things we were already doing, or needed to anyways. 

I also added a `renderWithContext()` helper for component tests that wraps `render()` from `@testing-library/react` for cases where some of our app's contexts are needed. I just added the react-query context for now, but we can add other necessary contexts for tests that need them as we bump into them.

* use `<Portal/>` to render it as a direct child of the body (not as a descendant of the app dom root).
* add various `aria-*` attributes to help screen readers
* Lock focus control to modal when it's visible
* Focus first tabbable element in the body automatically
* Prevent scrolling on `document.body` when modal is open so background doesn't scroll
* Make the `X` close icon a button so it's tabbable/focusable
* Clicking off the modal will close it now

## Related Issue(s)

Fixes WEB-393

## How to test

* Navigate around the Dashboard and try the various places modals are shown. Try tabbing through them.
* Try both desktop sized, and smaller screens like mobile

I ran through a test of the modals on the following browsers just to ensure there's no regressions.
- [x] Chrome (Mac)
- [x] Firefox (Mac)
- [x] Arc (Mac)
- [x] Brave (Mac)
- [x] Edge on Windows

The only thing noteworthy was a bug with Safari (iOS and desktop) where the footer buttons don't make it into the tab rotation for some reason. I'll look into this in a followup. The current state of tabbing in modals doesn't support footer buttons either, so fwiw this is no worse in that case.

## Documentation


#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - brad-web-3e7b20ecc3a</li>
	<li><b>🔗 URL</b> - <a href="https://brad-web-3e7b20ecc3a.preview.gitpod-dev.com/workspaces" target="_blank">brad-web-3e7b20ecc3a.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options:

      Run the build with werft instead of GHA
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

</details>

<details>
<summary>Installer Options</summary>

  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
      If enabled this will build `install/preview`
      If enabled this will create the environment on GCE infra
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold
